File: C/MergeSort.c
#include <stdio.h>
#include <stdlib.h>

void merge(int vet[], int begin, int mid, int end) { 
    int i, j, k; 
    int n1 = mid - begin + 1; 
    int n2 = end - mid; 

    int L[n1], R[n2]; 
    
    printf("L: ");
    for (i = 0; i < n1; i++) {
        L[i] = vet[begin + i];
        printf("%d ", L[i]);
    }
    
    printf("\nR: ");
    for (j = 0; j < n2; j++) {
        R[j] = vet[mid + 1+ j];
        printf("%d ", R[j]);
    }
    printf("\n");
  
    i = 0; 
    j = 0; 
    k = begin;
    while (i < n1 && j < n2) { 
        if (L[i] <= R[j]) { 
            vet[k] = L[i]; 
            i++; 
        } 
        else { 
            vet[k] = R[j]; 
            j++; 
        } 
        k++; 
    } 
  
    while (i < n1) { 
        vet[k] = L[i]; 
        i++; 
        k++; 
    } 

    while (j < n2) { 
        vet[k] = R[j]; 
        j++; 
        k++; 
    }
    
    for (i = 0; i < k; i++){
        printf("%d ", vet[i]);
    }
    printf("\n");
} 
  
void mergeSort(int vet[], int begin, int end) { 
    if (begin < end) { 
        int mid = (begin+end)/2; 
  
        mergeSort(vet, begin, mid); 
        mergeSort(vet, mid+1, end); 
  
        merge(vet, begin, mid, end); 
    } 
}

int main()
{
    int vet[7] = {8, 3, 2, 7, 1, 5, 4};
    
    int i, j, v, temp, menor;
    
    mergeSort(vet, 0, 6);
    
    for (i = 0; i < 7; i++) {
        printf("%d ", vet[i]);
    }

    return 0;
}

File: Python/MergeSort.py
def merge_sort(nums, start, end):
    if start < end:
        mid = (start + (end-1)) // 2
        merge_sort(nums, start, mid)
        merge_sort(nums, mid + 1, end)
        merge(nums, start, mid, end)


def merge(nums, start, mid, end):
    n1 = mid - start + 1
    n2 = end - mid

    left = [0] * (n1)
    right = [0] * (n2)

    for i in range(n1):
        left[i] = nums[start + i]
    for j in range(n2):
        right[j] = nums[mid + 1 + j]

    i, j = 0, 0
    k = start

    while i < n1 and j < n2:
        if left[i] <= right[j]:
            nums[k] = left[i]
            i += 1
        else:
            nums[k] = right[j]
            j += 1
        k += 1

    while i < n1:
        nums[k] = left[i]
        i += 1
        k += 1

    while j < n2:
        nums[k] = right[j]
        j += 1
        k += 1


if __name__ == "__main__":
    nums = [5, 2, 4, 6, 1, 3]
    merge_sort(nums, 0, len(nums) - 1)
    print(nums)
